docs/design-docs/03-pipeline-merge.md

```diff
@@ -0,0 +1,491 @@
+# Pipeline Merge
+
+* Author(s): Phani Teja Marupaka, phanimarupaka
+* Approver: Sunil Arora, Mike Borozdin
+
+## Why
+
+Currently, `kpt pkg update` doesn't merge the `pipeline` section in the Kptfile as expected. 
+The fact that the `pipeline` section is a non-associative list with no defined function identity makes 
+it very difficult to merge with upstream counterparts. Ordering of the functions is also important.
+This friction is forcing users to use `setters` and discouraging them from declaring other functions in the `pipeline` as 
+they will be deleted during the `kpt pkg update`. [Here](https://github.com/GoogleContainerTools/kpt/issues/2529) 
+is the example issue. Merging the pipeline correctly will reduce huge amounts of 
+friction in declaring new functions. This will encourage users to declare more functions 
+in the pipeline which in turn helps to **avoid excessive parameterization**.
+
+Consider the example of [Landing Zone](https://github.com/GoogleCloudPlatform/blueprints/tree/main/catalog) blueprints. 
+Parameters(setters) are the primary interface for the package. This is an anti-pattern for package-as-interface motivation 
+and one of the major blockers for not using other functions is the merge behavior 
+of the pipeline. If this problem is solved, LZ maintainers can rewrite the packages 
+with best practices aligned to the bigger goal of treating the package-as-interface, 
+discourage excessive use of setters and only use them as parameterization techniques of last resort.
+
+## Design
+
+In order to solve this problem, we should merge the pipeline section of the Kptfile 
+in a custom manner based on the most common use-cases and the expected outputs in 
+such scenarios. There are no user interface changes. Users will invoke the `kpt pkg 
+update` command in the same way they do currently. This effort will improve the merged 
+output of the pipeline section.
+
+**Is this change backwards compatible**: We are not making any changes to the api, 
+we are only improving the merged output of the `pipeline` section. This change will 
+produce a different output of Kptfile when compared to the current version but this 
+is not a breaking change.
+
+## User guide
+
+Here is what users can expect when they invoke the command, `kpt pkg update` on a package.
+
+#### Terminology
+
+**Original**: Source of the last fetched version of the package, represented by the upstreamLock section in Kptfile.
+
+**Updated upstream**: Declared source of the updated package, represented by the upstream section in Kptfile.
+
+**Local**: Local fork of the package on disk.
+
+Firstly, we need to define the identity of the function in order to uniquely identify 
+a function across three sources to perform a 3-way merge. In order to reliably identify the instance of a
+function, we should add a new optional field `name` to function definition.
+
+Here is the merge keys used by update logic to identify function in the order of precedence:
+
+1. name
```

Comment:
So is the idea here that if a function has the name field specified, we will use that for merge. And for functions that doesn't have a name, we will use the other properties? And does the order only come in to break ties, or do we always consider that as part of the merge key?

I'm a bit worried that using so many fields in the merge key will make it hard to explain and predict what the result of a merge will look like. In particular, considering the ordering can make for surprising behavior.

Could we make this simpler by using the following approach:
- If all functions in upstream, local and origin has a name, merge using the name as the merge key.
- If none of the functions in upstream, local, and origin has a name, merge as a non-associative list just like now.
- If only some of the functions have a name, print a warning to users and merge as a non-associative list.

I think this keeps current behavior if users doesn't make changes, but provides a clear path to performing the merge as an associative list. It has only two behaviors, both of which are pretty easy to understand.

Reply:
`So is the idea here that if a function has the name field specified, we will use that for merge. And for functions that doesn't have a name, we will use the other properties? And does the order only come in to break ties, or do we always consider that as part of the merge key? ` -> Yes. order only come in to break ties.

The aim here is to not burden users by forcing them to add `name` field. Also there will be a migration cost for adding name fields to existing upstream functions. Using `image` name as the merge-key will solve the problem for 95% of the use-cases. For the complicated cases(function declared multiple times), users can choose to add name field. 

The current merge behavior is very undesirable and users are already running into issues. This change will make the merge behavior better with no migration cost to the users.